### PR TITLE
helmchart apply ordering

### PR DESCRIPTION
**Pull request by laverya**
I'm honestly kind of miffed with the whole "toss a magic comment at the beginning of the helm chart file" solution, but I didn't find a better one.

#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note
Added a 'weight' parameter to native helm charts. Higher weights will be applied first.
```

#### Does this PR require documentation?

Yes, there will need to be a new section added to https://docs.replicated.com/reference/custom-resource-helmchart.

### CODE REVIEW

**Review by emosbaugh — APPROVED** [review]

File: pkg/operator/client/deploy.go [hunk 1]
```diff
@@ -506,33 +509,26 @@ func (c *Client) clearNamespace(slug string, namespace string, isRestore bool, r
...
+
 func (c *Client) installWithHelm(helmDir string, targetNamespace string) (*commandResult, error) {
 	version := "3"
 	chartsDir := filepath.Join(helmDir, "charts")
-	dirs, err := ioutil.ReadDir(chartsDir)
+
+	orderedDirs, err := getSortedCharts(chartsDir)
 	if err != nil {
-		return nil, errors.Wrap(err, "failed to read archive dir")
+		return nil, err
```
[the remark this answers is not in the document]
Comment (laverya) [reply pkg/operator/client/deploy.go:524]: I can

File: pkg/operator/client/deploy.go [hunk 2]
```diff
@@ -564,6 +560,69 @@ func (c *Client) installWithHelm(helmDir string, targetNamespace string) (*comma
...
+
+		reader := bufio.NewReader(bytes.NewBuffer(chartFile))
+		firstLineBytes, _, err := reader.ReadLine()
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to read first line of %s", chartfilePath)
+		}
+		firstLine := string(firstLineBytes)
+
+		// get the weight from the first line
+		if !strings.HasPrefix(firstLine, "#kotsweight: ") { // TODO make '#kotsweight' a constant
```
[the remark this answers is not in the document]
Comment (laverya) [reply pkg/operator/client/deploy.go:594]: I'll move it to a constant

**Review by sgalsaleh — COMMENTED** [review]
kots kinds schema need to be regenerated

File: pkg/operator/client/deploy.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (laverya) [reply pkg/operator/client/deploy.go]: this has been removed entirely instead

File: pkg/base/replicated.go [hunk 3]
```diff
@@ -233,7 +233,7 @@ func renderReplicatedHelmChart(kotsHelmChart *kotsv1beta1.HelmChart, upstreamFil
 		ExcludeKotsKinds:  renderOptions.ExcludeKotsKinds,
 		Log:               nil,
 		UseHelmInstall:    kotsHelmChart.Spec.UseHelmInstall,
-	})
+	}, kotsHelmChart.Spec.Weight)
```
[the remark this answers is not in the document]
Comment (laverya) [reply pkg/base/replicated.go]: it's been removed now

File: pkg/base/replicated_test.go [hunk 4]
```diff
@@ -690,3 +691,564 @@ func Test_removeFileFromUpstream(t *testing.T) {
...
+    - name: tcp-postgresql
+      port: 5432
+      targetPort: tcp-postgresql
+  selector:
+    app.kubernetes.io/name: postgresql
+    app.kubernetes.io/instance: postgresql`),
+						},
+						{
+							Path: "svc.yaml",
+							Content: []byte(`# Source: postgresql/templates/svc.yaml
```
[the remark this answers is not in the document]
Comment (laverya) [reply pkg/base/replicated_test.go:1134]: I have no clue! 

Ended up not changing the code that this tests at all, but I figured I'd leave in the unit test after writing it.

**Review by sgalsaleh — APPROVED** [review]

**Review by sgalsaleh — APPROVED** [review]

File: pkg/base/replicated_test.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (emosbaugh) [reply pkg/base/replicated_test.go:1134]: helm template

```
✗ helm template prometheus-community/kube-prometheus-stack --kube-version  1.22 | grep '# Source'
# Source: kube-prometheus-stack/charts/grafana/templates/serviceaccount.yaml
# Source: kube-prometheus-stack/charts/grafana/templates/tests/test-serviceaccount.yaml
# Source: kube-prometheus-stack/charts/kube-state-metrics/templates/serviceaccount.yaml
# Source: kube-prometheus-stack/charts/prometheus-node-exporter/templates/serviceaccount.yaml
# Source: kube-prometheus-stack/templates/alertmanager/serviceaccount.yaml
# Source: kube-prometheus-stack/templates/prometheus-operator/serviceaccount.yaml
...
```

**Pull request merged**